spec/PeriodicBackgroundSync-index.bs

```diff
@@ -92,3 +94,95 @@ Fetch requests within the <code>periodicsync</code> event while [=in the backgro
 
 ## History Leaking ## {#history-leaking}
 Fetch requests within the <code>periodicsync</code> event while [=in the background=] may reveal something about the client's navigation history to middleboxes on networks different from the one used to create the [=periodicsync registration=]. For instance, the client might visit site https://example.com, which registers a <code>periodicsync</code> event, but based on the implementation, might not fire until after the user has navigated away from the page and changed networks. Middleboxes on the new network may see the fetch requests that the <code>periodicsync</code> event makes. The fetch requests are HTTPS so the request contents will not be leaked but the domain may be (via DNS lookups and IP address of the request). To prevent this leakage of browsing history, the user agent MAY choose to only fire <code>periodicsync</code> events on the network the [=periodicsync registration=] was made on, with the understanding that it will reduce usability by not allowing synchronization opportunistically.
+
+Algorithms {#algorithms}
+=========================
+## <dfn>Calculate a time to fire</dfn> ## {#caculate-time-to-fire}
+This section describes how a user agent can calculate the [=time to fire=] for a [=periodicsync registration=], <var>registration</var>. The time interval between periodicSync events for |registration| MUST be greater than or equal to the [=minInterval=] value associated with |registration|.
+
+The user agent MAY include factors such as user engagement with the origin to decide this time interval, allowing origins with high user engagement to update their web apps more often. In addition, the user agent SHOULD have per-origin and cross-origin caps on this interval, as recommended in [[#privacy]] and Resource Usage sections.
+
+The user agent may also decide to retry each failed periodicSync event, capped to a user agent defined <dfn>maximum number of retries</dfn>.
+
+A possible algorithm to calculate the [=time to fire=], <var>timeToFire</var> for |registration| would involve running these steps:
+1. If this is the <dfn>first attempt of firing</dfn> the periodicSync event:
+    1. Let <var>now</var> represent the current time.
+    1. Let <var>siteEngagementLevel</var> represent the level of engagement the user has with the origin. Calculate <var>siteEngagementPenalty</var>, which is a factor of |siteEngagementLevel|, and is inversely proportional to it. Let <var>minPeriodicSyncInterval</var> be a global cap on time intervals between periodicSync events. Define <var>effectiveGap</var> as |siteEngagementPenalty|*|minPeriodicSyncInterval|.
```

Comment:
I'd also mention how to get the origin. Either pass it as a function parameter, or get it from the SW registration. Preferably pass it as a parameter, so the main register function can do all of the state assertions first.